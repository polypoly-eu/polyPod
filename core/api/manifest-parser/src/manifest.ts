import * as Decode from "io-ts/lib/Decoder";
import { fold } from "fp-ts/lib/Either";
import { pipe } from "fp-ts/lib/pipeable";
import { parse as parseSemVer, SemVer, Range } from "semver";

export interface MainManifest {
    readonly name: string;
    readonly version: SemVer;
}

export interface FeatureManifest {
    readonly description: string;
    readonly thumbnail: string;
    readonly primaryColor: string;
    // TODO: Typecheck links object
    readonly links: unknown;
    // TODO: Typecheck translations object
    readonly translations: unknown;
}

export interface Manifest extends MainManifest, FeatureManifest {}

// TODO duplicated code with podigree, should be a library
function expect<I, A>(input: I, msg: string, decoder: Decode.Decoder<I, A>): A {
    return pipe(
        decoder.decode(input),
        fold(
            (error) => {
                throw new Error(msg + "\n" + Decode.draw(error));
            },
            (t) => t
        )
    );
}

const relativeDecoder = pipe(
    Decode.string,
    Decode.parse((string) => {
        if (typeof document == "undefined") {
            return Decode.success(string);
        }
        const url = new URL(string, document.location.href);
        if (url.toString() == "") return Decode.failure(string, "relative");

        return Decode.success(string);
    })
);

const mainDecoder = Decode.type({
    name: Decode.string,
    version: pipe(
        Decode.string,
        Decode.parse((string) => {
            const parsed = parseSemVer(string);
            if (parsed === null) return Decode.failure(string, "version string");

            return Decode.success(parsed);
        })
    ),
});

const featureDecoder = Decode.type({
    name: Decode.string,
    description: Decode.string,
    thumbnail: relativeDecoder,
    primaryColor: Decode.string,
    links: Decode.UnknownRecord,
    translations: Decode.UnknownRecord,
});

export async function readManifest(packageManifest: Record<string, unknown>): Promise<Manifest> {
    const rawMain = expect(packageManifest, "Failed to parse main manifest", mainDecoder);
    const featureManifest = expect(
        packageManifest,
        "Failed to parse Feature manifest",
        featureDecoder
    );

    return {
        name: featureManifest.name,
        version: rawMain.version,
        description: featureManifest.description,
        thumbnail: featureManifest.thumbnail,
        primaryColor: featureManifest.primaryColor,
        links: featureManifest.links,
        translations: featureManifest.translations,
    };
}
