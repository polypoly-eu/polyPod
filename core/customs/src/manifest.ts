import * as Decode from "io-ts/lib/Decoder";
import { fold } from "fp-ts/lib/Either";
import readPkg from "@pnpm/read-package-json";
import { pipe } from "fp-ts/lib/pipeable";
import { parse as parseSemVer, SemVer, Range } from "semver";
import { normalize, isAbsolute, join, dirname } from "path";
import { promises as fs } from "fs";
import { Url } from "url";
import { unknown } from "io-ts";

export interface EngineManifest {
    readonly api: Range;
}

export interface MainManifest {
    readonly name: string;
    readonly version: SemVer;
}

export interface RootManifest {
    readonly root: string;
}

export interface FeatureManifest {
    readonly name: string;
    readonly description: string;
    readonly thumbnail: string;
    readonly primaryColor: string;
    // TODO: Typecheck links object
    // readonly links: Map<string, Url>;
    readonly links: unknown[];
    // TODO: Typecheck translations object
    //readonly translations: Array<Translation>;
    readonly translations: unknown[];
}

export interface Manifest extends EngineManifest, MainManifest, RootManifest, FeatureManifest {}

export interface Link {
    readonly name: string;
    readonly URL: Url;
}

// TODO duplicated code with podigree, should be a library
function expect<I, A>(input: I, msg: string, decoder: Decode.Decoder<I, A>): A {
    return pipe(
        decoder.decode(input),
        fold(
            (error) => {
                throw new Error(msg + "\n" + Decode.draw(error));
            },
            (t) => t
        )
    );
}

const relativeDecoder = pipe(
    Decode.string,
    Decode.parse((string) => {
        const path = normalize(string);

        if (isAbsolute(path) || path.startsWith("..")) return Decode.failure(string, "relative");

        return Decode.success(string);
    })
);

const mainDecoder = Decode.type({
    name: Decode.string,
    version: pipe(
        Decode.string,
        Decode.parse((string) => {
            const parsed = parseSemVer(string);
            if (parsed === null) return Decode.failure(string, "version string");

            return Decode.success(parsed);
        })
    ),
});

const engineDecoder = Decode.type({
    polypoly: pipe(
        Decode.string,
        Decode.parse((string) => {
            try {
                return Decode.success(new Range(string));
            } catch (err) {
                return Decode.failure(string, err.message);
            }
        })
    ),
});

const rootDecoder = Decode.type({
    polypoly: Decode.type({
        root: relativeDecoder,
        manifest: relativeDecoder,
    }),
});

const featureDecoder = Decode.type({
    name: Decode.string,
    description: Decode.string,
    thumbnail: relativeDecoder,
    primaryColor: Decode.string,
    links: Decode.UnknownArray,
    translations: Decode.UnknownArray,
});

export async function readManifest(pkgPath: string): Promise<Manifest> {
    const packageManifest = await readPkg(pkgPath);

    const rawMain = expect(packageManifest, "Failed to parse main manifest", mainDecoder);
    const rawEngine = expect(packageManifest.engines, "Failed to parse engines", engineDecoder);
    const rawRoot = expect(packageManifest, "Failed to parse Feature spec", rootDecoder);

    let featureManifest: FeatureManifest = {
        name: rawMain.name,
        description: "",
        thumbnail: "",
        primaryColor: "",
        links: [],
        translations: [],
    };

    if (rawRoot.polypoly.manifest) {
        const manifestPath = join(dirname(pkgPath), rawRoot.polypoly.manifest);
        const featureManifestJson = JSON.parse(await fs.readFile(manifestPath, "utf8"));

        featureManifest = expect(
            featureManifestJson,
            "Failed to parse Feature manifest",
            featureDecoder
        );
    }

    return {
        api: rawEngine.polypoly,
        root: rawRoot.polypoly.root,
        name: featureManifest.name,
        version: rawMain.version,
        description: featureManifest.description,
        thumbnail: featureManifest.thumbnail,
        primaryColor: featureManifest.primaryColor,
        links: featureManifest.links,
        translations: featureManifest.translations,
    };
}
