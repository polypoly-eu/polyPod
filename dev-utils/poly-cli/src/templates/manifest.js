export function manifestTemplate(feature_name, author, version, feature_type) {
    let base = {
        name: feature_name,
        author: author,
        version: version,
        description: "",
        thumbnail: "",
        thumbnailColor: "#ffffff",
        primaryColor: "#ffffff",
        links: {},
        translations: {
            de: {
                author: "",
                description: "",
            },
        },
    };

    if (feature_type === "preview") {
        base.links = {
            "learn-more": "https://polypoly.org",
        };
    }

    return JSON.stringify(base, null, 2);
}
