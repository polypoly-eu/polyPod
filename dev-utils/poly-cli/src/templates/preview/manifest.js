export function manifestTemplate(feature_name, author) {
    let t = {
        name: feature_name,
        author: author,
        description: "",
        thumbnail: "",
        primaryColor: "#ffffff",
        links: {
            "learn-more": "https://polypoly.org",
        },
        translations: {
            de: {
                author: "",
                description: "",
                links: {},
            },
        },
    };

    return JSON.stringify(t, null, 2);
}
