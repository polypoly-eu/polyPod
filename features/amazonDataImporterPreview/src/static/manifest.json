{
  "name": "amazon-data-importer-preview",
  "author": "polypoly poly-cli",
  "version": "0.0.1",
  "description": "",
  "thumbnail": "",
  "thumbnailColor": "#ffffff",
  "primaryColor": "#0f1938",
  "borderColor": "#cbd5e0",
  "links": {
    "learn-more": "https://polypoly.org"
  },
  "translations": {
    "de": {
      "author": "",
      "description": ""
    }
  }
}