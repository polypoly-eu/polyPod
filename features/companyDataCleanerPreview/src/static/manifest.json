{
  "name": "Company Data Cleaner",
  "author": "polypoly poly-cli",
  "version": "0.0.1",
  "description": "Correcting your data forces companies to spend a lot of money – money that you can earn.",
  "thumbnail": "images/thumbnail.pdf",
  "thumbnailColor": "#ffffff",
  "primaryColor": "#0f1938",
  "borderColor": "#cbd5e0",
  "links": {
    "learn-more": "https://polypoly.org"
  },
  "translations": {
    "de": {
      "author": "",
      "description": "Die Korrektur Ihrer Daten zwingt Unternehmen, viel Geld auszugeben – Geld, das Sie verdienen können."
    }
  },
  "tileTextColor": "#0f1938"
}