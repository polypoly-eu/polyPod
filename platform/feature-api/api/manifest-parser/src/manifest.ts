import * as Decode from "io-ts/lib/Decoder";
import { fold } from "fp-ts/lib/Either";
import { pipe } from "fp-ts/function";
import { parse as parseSemVer, SemVer, Range } from "semver";

export interface MainManifest {
    readonly name: string;
    readonly version: SemVer;
}

export interface FeatureManifest {
    readonly description: string;
    readonly thumbnail: string;
    readonly primaryColor: string;
    // TODO: Typecheck links object
    readonly links: unknown;
    // TODO: Typecheck translations object
    readonly translations: unknown;
}

export interface Manifest extends MainManifest, FeatureManifest {}

function expect<I, A>(input: I, msg: string, decoder: Decode.Decoder<I, A>): A {
    return pipe(
        decoder.decode(input),
        fold(
            (error) => {
                throw new Error(msg + "\n" + Decode.draw(error));
            },
            (t) => t
        )
    );
}

const relativeDecoder = pipe(
    Decode.string,
    Decode.parse((string) => {
        if (typeof document == "undefined") {
            return Decode.success(string);
        }
        const url = new URL(string, document.location.href);
        if (url.toString() == "") return Decode.failure(string, "relative");

        return Decode.success(string);
    })
);

const mainDecoder = Decode.type({
    name: Decode.string,
    version: pipe(
        Decode.string,
        Decode.parse((string) => {
            const parsed = parseSemVer(string);
            if (parsed === null) return Decode.failure(string, "version string");

            return Decode.success(parsed);
        })
    ),
});

const featureDecoder = Decode.type({
    name: Decode.string,
    description: Decode.string,
    thumbnail: relativeDecoder,
    primaryColor: Decode.string,
    links: Decode.UnknownRecord,
    translations: Decode.UnknownRecord,
});

export async function readManifest(packageManifest: Record<string, unknown>): Promise<Manifest> {
    // There is no 'version' property in real feature manifests at this time, but to make this code
    // happy, we're simply adding one if it's missing.
    const manifestWithVersion = { ...packageManifest };
    if (!("version" in manifestWithVersion)) manifestWithVersion.version = "1.0.0";

    const rawMain = expect(manifestWithVersion, "Failed to parse main manifest", mainDecoder);
    const featureManifest = expect(
        manifestWithVersion,
        "Failed to parse Feature manifest",
        featureDecoder
    );

    return {
        name: featureManifest.name,
        version: rawMain.version,
        description: featureManifest.description,
        thumbnail: featureManifest.thumbnail,
        primaryColor: featureManifest.primaryColor,
        links: featureManifest.links,
        translations: featureManifest.translations,
    };
}
