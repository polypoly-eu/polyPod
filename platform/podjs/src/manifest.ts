import * as Decoder from "io-ts/Decoder";
import * as Either from "fp-ts/Either";
import { pipe } from "fp-ts/function";

export interface Manifest {
    readonly name: string;
    readonly description: string;
    readonly thumbnail: string;
    readonly primaryColor: string;
    readonly borderColor: string;
    readonly borderSize: string;
    readonly links: Record<string, string>;
    readonly translations: Record<string, Partial<Manifest>>;
}

/**
 * It takes a decoder and an input, and returns the output of the decoder
 * @param {DecodeFrom} input - The manifest to be parsed.
 * @param decoder - Decoder.Decoder<DecodeFrom, EncodeTo>
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const decodeWith = <EncodeTo = any, DecodeFrom = unknown>(
    input: DecodeFrom,
    decoder: Decoder.Decoder<DecodeFrom, EncodeTo>
): EncodeTo =>
    pipe(
        decoder.decode(input),
        Either.getOrElseW((errors) => {
            throw new Error("Failed to parse manifest: " + errors);
        })
    );

/* It's a decoder that parses a string to a URL. Fails on error inputs (i.e. relative paths) */
const relativeDecoder = pipe(
    Decoder.string,
    Decoder.parse((input) => {
        if (typeof document == "undefined") {
            return Decoder.success(input);
        }
        const url = new URL(input, document.location.href);
        if (url.toString() == "") {
            return Decoder.failure(input, "relative");
        }
        return Decoder.success(input);
    })
);

/* It is a type alias for a decoder that would decode a manifest */
const manifestDecoder = Decoder.type({
    name: Decoder.string,
    description: Decoder.string,
    thumbnail: relativeDecoder,
    primaryColor: Decoder.string,
    borderColor: Decoder.string,
    borderSize: Decoder.string,
    links: Decoder.record(Decoder.string),
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    translations: Decoder.record(Decoder.partial<any>({})),
});

/**
 * It reads the manifest that is passed as a json format and returns it as a Manifest object.
 * @param packageManifest - The manifest of the package.
 * @returns A `Manifest` object.
 */
export async function readManifest(
    packageManifest: Record<string, unknown>
): Promise<Manifest> {
    const manifest = decodeWith(packageManifest, manifestDecoder);

    return {
        name: manifest.name,
        description: manifest.description,
        thumbnail: manifest.thumbnail,
        primaryColor: manifest.primaryColor,
        links: manifest.links,
        translations: manifest.translations,
        borderColor: manifest.borderColor,
        borderSize: manifest.borderSize,
    };
}
