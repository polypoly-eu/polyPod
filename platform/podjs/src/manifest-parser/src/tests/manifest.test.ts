import { readManifest } from "../manifest";
import { join } from "path";

describe("Parsing", () => {
    describe("Successful parse", () => {
        it.each(["manifest.json"])("%s", async (pkg) => {
            expect(
                await readManifest(await import(join(__dirname, "data", pkg)))
            ).toMatchSnapshot();
        });
    });

    describe("Failed parse", () => {
        it.each(["fail-empty.json"])("%s", async (pkg) => {
            await expect(
                readManifest(await import(join(__dirname, "data", pkg)))
            ).rejects.toThrowErrorMatchingSnapshot();
        });
    });
});
