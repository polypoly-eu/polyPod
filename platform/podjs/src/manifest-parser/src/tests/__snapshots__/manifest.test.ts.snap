// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parsing Failed parse fail-empty.json 1`] = `
"Failed to parse main manifest
required property \\"name\\"
└─ cannot decode undefined, should be string"
`;

exports[`Parsing Successful parse manifest.json 1`] = `
Object {
  "description": "This is an example feature",
  "links": Object {
    "link1": "https://polypoly.coop",
    "link2": "https://polypoly.coop",
  },
  "name": "The Test Feature",
  "primaryColor": "#000000",
  "thumbnail": "assets/thumbnail.png",
  "translations": Object {
    "de": Object {
      "author": "Author_DE",
      "description": "Description_DE",
    },
  },
}
`;
